consensus/safety-rules/src/logging.rs

```diff
@@ -1,14 +1,105 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-use libra_logger::StructuredLogEntry;
+use consensus_types::common::Round;
+use libra_logger::{Key, Schema, StructuredLogEntry, Value, Visitor};
+use libra_types::waypoint::Waypoint;
+use serde::Serialize;
 
-pub fn safety_log(entry: LogEntry, event: LogEvent) -> StructuredLogEntry {
-    StructuredLogEntry::new_named("safety_rules", entry.as_str())
-        .data(LogField::Event.as_str(), event.as_str())
+// TODO implement a proc macro which can generate all of the boilerplate for this struct
+pub struct SafetyLogSchema {
+    name: LogEntry,
+    event: LogEvent,
+    round: Option<Round>,
+    preferred_round: Option<u64>,
+    last_voted_round: Option<u64>,
+    epoch: Option<u64>,
+    error: Option<String>,
+    waypoint: Option<Waypoint>,
```

Comment:
This is the key part to look at. We can have folks define schemas for their logs and then we can generate the boiler plate below, including all of the setters and the "new" function

Reply:
@davidiw Ideally the "error" field would actually hold a `&'a Error` instead of a string, but I ran into some difficulties with some of the lifetimes that it introduced in some "callback" code in safety_rules.rs